beacon-chain/operations/slashings/service.go

```diff
@@ -0,0 +1,194 @@
+	pending := make([]*ethpb.ProposerSlashing, 0, params.BeaconConfig().MaxProposerSlashings)
+	for i, slashing := range p.pendingProposerSlashing {
+		if i >= int(params.BeaconConfig().MaxProposerSlashings) {
+			break
+		}
+		pending = append(pending, slashing)
+	}
+	return pending
+}
+
+// InsertAttesterSlashing into the pool. This method is a no-op if the attester slashing already exists in the pool,
+// has been included into a block recently, or the validator is already exited.
+func (p *Pool) InsertAttesterSlashing(state *beaconstate.BeaconState, slashing *ethpb.AttesterSlashing) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	slashedVal := sliceutil.IntersectionUint64(slashing.Attestation_1.AttestingIndices, slashing.Attestation_2.AttestingIndices)
+	for _, val := range slashedVal {
+		// Has this validator index been included recently?
+		if p.included[val] {
+			continue
+		}
+		validator, err := state.ValidatorAtIndexReadOnly(val)
+		if err != nil {
+			return err
+		}
+		// Has the validator been exited already?
+		if validator.ExitEpoch() < helpers.CurrentEpoch(state) {
+			continue
+		}
+		//Has the validator been slashed already?
+		slashedValidators := state.Slashings()
+		found := sort.Search(len(slashedValidators), func(i int) bool {
+			return slashedValidators[i] == val
+		})
+		if found != len(slashedValidators) {
+			continue
+		}
+
+		// Does this validator exist in the pending list already? Use binary search to find the answer.
+		found = sort.Search(len(p.pendingAttesterSlashing), func(i int) bool {
+			return p.pendingAttesterSlashing[i].validatorToSlash == val
+		})
+		if found != len(p.pendingAttesterSlashing) {
+			continue
+		}
+
+		pendingSlashing := &PendingAttesterSlashing{
+			attesterSlashing: slashing,
+			validatorToSlash: val,
+		}
+
+		// Insert into pending list and sort again.
+		p.pendingAttesterSlashing = append(p.pendingAttesterSlashing, pendingSlashing)
+		sort.Slice(p.pendingAttesterSlashing, func(i, j int) bool {
+			return p.pendingAttesterSlashing[i].validatorToSlash < p.pendingAttesterSlashing[j].validatorToSlash
+		})
+	}
+	return nil
+}
+
+// InsertProposerSlashing into the pool. This method is a no-op if the pending slashing already exists,
+// has been included recently, the validator is already exited, or the validator was already slashed.
+func (p *Pool) InsertProposerSlashing(state *beaconstate.BeaconState, slashing *ethpb.ProposerSlashing) error {
+	p.lock.Lock()
+	defer p.lock.Unlock()
+	idx := slashing.ProposerIndex
+
+	// Has this validator index been included recently?
+	if p.included[idx] {
+		return nil
+	}
+
+	validator, err := state.ValidatorAtIndexReadOnly(idx)
+	if err != nil {
+		return err
+	}
+	// Has the validator been exited already?
+	if validator.ExitEpoch() < helpers.CurrentEpoch(state) {
+		return nil
+	}
+
+	// Has the validator been slashed already?
+	slashedValidators := state.Slashings()
+	found := sort.Search(len(slashedValidators), func(i int) bool {
```

Comment:
I see duplicated code here and in the func above, could we refactor for deduplication or use a helper?

Reply:
Resolved